optimum/onnxruntime/trainer.py

```diff
@@ -279,12 +308,48 @@ def __init__(
             preprocess_logits_for_metrics=preprocess_logits_for_metrics,
         )
 
+        # We leverage both training_model and inference_model in conjunction with model.
+        # _training_model will be wrapped so it will use ORT and will use the overriden functions in ModuleWithLoss.
+        # _inferencing_model will be storing the default version of the model and we will switch to it in case of eval/test.
+
+        # Only Wrap the model if we pass --loss_in_train flag.
+        if args.loss_in_train:
+            self._training_model = ModuleWithLoss(model, args)
```

Comment:
nit: maybe we can pass `self` into ModuleWithLoss int its constructor. Then you can get self.model, self.args, and even self.label_smoothers inside the ModuleWithLoss class.

Reply:
here is an example:

```
import types


def create_model_with_loss(self):
    class ModuleWithLoss(nn.Module):
        def __init__(self, model, args, label_smoother):
            super().__init__()
            self._original_model = model
            self.args = args
            self.label_smoother = label_smoother

        def forward(self, inputs: Dict[str, Union[torch.Tensor, Any]], return_outputs):
            # The compute_model_plus_loss_internal is assigned once the class is instantiated.
            # It should have same signature as Trainer.compute_loss()
            return self.compute_model_plus_loss_internal(self._original_model, inputs, return_outputs)

        @property
        def config(self):
            return self._original_model.config

    model_with_loss = ModuleWithLoss(self.model, self.args, self.label_smoother)
    model_with_loss.compute_model_plus_loss_internal = types.MethodType(self.super().compute_loss, model_with_loss)

    return model_with_loss
```